src/main/java/com/google/sps/Scheduler.java

```diff
@@ -1,15 +1,134 @@
+import java.util.stream.Collectors;
 
 /**
  * Static class for permuting schedules given a list of courses.
  */
 public final class Scheduler {
-    public Collection<Schedule> generateSchedules(Collection<Course> courses, Invariants invariants){
-        //TODO: This
-        return null;
+  /**
+   * Generates all valid schedules given a list of courses and invariant
+   * constraints.
+   * 
+   * @param courses    The list of courses to generate the schedules from.
+   * @param invariants The invariants that constrain the valid schedules
+   * @return A list of valid schedules.
+   */
+  public List<Schedule> generateSchedules(List<Course> courses, Invariants invariants) {
+    List<Course> requiredCourses = new ArrayList<>();
+    List<Course> nonRequiredCourses = new ArrayList<>();
+
+    for (Course course: courses) {
+      if (course.isRequired()) {
+        requiredCourses.add(course);
+      } else {
+        nonRequiredCourses.add(course);
+      }
+    }
+
+    // Generate all possible valid Schedules that include only required courses.
+    float reqCredits = totalCredits(requiredCourses);
+    Set<List<Course>> requiredSet = new HashSet<>();
+
+    generateSchedulesHelper(requiredCourses, new ArrayList<>(), 
+        new Invariants(reqCredits, reqCredits), requiredSet);
+    
+    // Using those Schedules as a building block, generate all valid schedules
+    Set<List<Course>> scheduleSet = new HashSet<>();
+    for (List<Course> validCourseList : requiredSet) {
+      generateSchedulesHelper(nonRequiredCourses, validCourseList, invariants, scheduleSet);
+    }
+
+    List<Schedule> schedules = new ArrayList<>();
+    for (List<Course> courseList: scheduleSet) {
+      schedules.add(new Schedule(courseList));
+    }
+
+    return schedules;
+  }
+
+  /**
+   * Recursively iterates through a list of courses and adds schedules to the
+   * given set.
+   * 
+   * @param courses     The list of courses to iterate through
+   * @param courseList  The current course list to be added/operated on
+   * @param invariants  The invariants that constrain the valid schedules
+   * @param courseLists The set of course lists that is being generated/added to.
+   */
+  private void generateSchedulesHelper(List<Course> courses, List<Course> courseList, 
+      Invariants invariants, Set<List<Course>> courseLists) {
+
+    // This means that we have a valid schedule, and should add it to the list.
+    float currCredits = totalCredits(courseList);
+    if (invariants.meetsCreditRequirement(currCredits)) {
+      courseLists.add(courseList);
+    }
+
+    // We've reached the end of the list or have a full schedule, we are finished.
+    if (courses.isEmpty() || currCredits >= invariants.getMaxCredits()) {
+      return;
+    }
+
+    Course course = courses.remove(0);
+    List<Section> courseSections = course.getSections();
+    List<Section> scheduleSections = courseList.stream()
+        .map(c -> c.getSections().get(0))
```

Comment:
This relies on the assumption that each course has exactly one section.

If you want to get the list of all schedule sections, including the courses that have zero sections or more than one sections, you can use `Stream::flatMap()` instead:

```
    List<Section> scheduleSections = courseList.stream()
        .flatMap(c -> c.getSections().stream())
        .collect(Collectors.toList());
```

This may or may not be what you want though. Depending on how you handle this, you could potentially end up trying to add two sections of the same course to a schedule.